### PR TITLE
Implement HTTP 1.0 & fix TCP_client

### PR DESCRIPTION
**Pull request by gpotter2**
This PR implements:
- **HTTP 1.0 layer**. Based on `scapy_http` but:
    - added sessions support: follow chunks/streams of TCP HTTP (Content_Length and Chunked)
    - added decompression & re-compression support (3 default methods implemented: gzip, lzip, zlib)
    - fixes build: quite a few things were broken
    - add support for all HTTP fields
    - fixed for PEP8
    - improved Python 3 support, general bytes handling (especially with regexes/split)
    - Removed deprecated fields (Let's not import already-deprecated stuff)
    - added tests: test chunk reassembly, content_length threshold, TCPSession, gzip decompression (most common)
- **TCPsessions**. Used to sort TCP packets via their seq/ack. Used to decode HTTP flows. Could be extended to other TCP protocols
- **`TCP_client` & Automaton**: fixes for Automaton: support `_ATMT_Supersocket` on Windows. Also update it to h'share the `StringBuffer` of the `TCPSession`, so that it can also handle TCP fragmentation.
- **documentation** about HTTP/TCP_client
- add other specific tests (gzip compact utiles...)

Note: because of Python 2-3 compatibility, several utils were needed:
- `gzip_decompress/gzip_compress`: `gzip.decompress/gzip.compress` isn't available on Python 2.7.

~~**Important note: `sr()` is NOT supported YET, because it would require to change `sndrcv()` to handle Sessions. This is done in https://github.com/secdev/scapy/pull/1999.** (The 2 PRs must be merged before it may be implemented)~~ (done)

This code was able to decode a JPEG image sent through several HTTP chunks, despite TCP re-transmission. You can check it out by yourself by un-commenting a line in the chunked-image test

#### **Rationale**

- Why `scapy_http` ?

It has been around [for quite some time](https://github.com/invernizzi/scapy-http/commit/d716bdff62f0c46253319487ae9e31bb79b5f28b). It allows users/projects that were using it to migrate seamlessly, to use the new features.
I do not want to change the API, and it allowed to start on an existing basis.

- What's next ?

This PR ***should*** be able to fully replace [`scapy_http`](https://github.com/invernizzi/scapy-http) which development [has been paused (stopped?)](https://github.com/invernizzi/scapy-http/commit/1f9bde87e4e47f5fb65e2a341f4388b3fe39a239). That would allow to deprecate `scapy_http` and continue any further development here.
I've asked the maintainers if they want to merge the projects, therefore release a last `scapy_http` version with a warning deprecating it. I doubt they want to invest more time into maintaining it 😄 

**See https://github.com/invernizzi/scapy-http/issues/46**

### CODE REVIEW

File: scapy/sessions.py [hunk 1]
```diff
@@ -75,3 +75,36 @@ def _ip_process_packet(self, packet):
...
+               '{IP:%IP.dst%}{IPv6:%IPv6.dst%}:%r,TCP.dport%')
+        ack_fmt = ('TCP {IP:%IP.dst%}{IPv6:%IPv6.dst%}:%r,TCP.dport% > ' +
+                   '{IP:%IP.src%}{IPv6:%IPv6.src%}:%r,TCP.sport%')
+        seq = pkt[TCP].seq
+        ident = pkt.sprintf(fmt)
+        self.tcp_frags[ident][0][pkt] = seq
+        # Reassemble using all previous packets
+        print(seq)
+        print(":".join(str(x) for x in self.tcp_frags[ident][0].keys()))
+        return pay_class.tcp_reassemble(*self.tcp_frags[ident])
```
Comment (gpotter2) [inline scapy/sessions.py]: Not working. Still need to think this through

Comment (gpotter2) [reply scapy/sessions.py]: Updated.

**Comment by gpotter2** [comment]
@guedou @p-l- I finally finished this PR. Feel free to review it 😄

File: scapy/contrib/gtp.uts [hunk 2]
```diff
@@ -23,9 +23,10 @@ gtp.dport == 2123 and gtp.teid == 2807 and len(gtp.IE_list) == 5
 = GTPCreatePDPContextRequest(), basic dissection
 ~ random_weird_py3
 random.seed(0x2807)
-raw(gtp)
-assert _ in [b"E\x00\x00K\x00\x01\x00\x00@\x11|\x9f\x7f\x00\x00\x01\x7f\x00\x00\x01\x08K\x08K\x007\x1a\xc30\x10\x00'\x00\x00\n\xf7\x10\xd6\xd2\xf6\xd8\x14\x0b\x85\x00\x04\xa3\xad\x98\xfa\x85\x00\x04F\\`\xd6\x87\x00\x0fBCD5lPP8N2u8h9l",
-             b"E\x00\x00K\x00\x01\x00\x00@\x11|\x9f\x7f\x00\x00\x01\x7f\x00\x00\x01\x08K\x08K\x007\xd3\xf20\x10\x00'\x00\x00\n\xf7\x107m\xaeV\x14\x08\x85\x00\x04\x83\x92K\xa3\x85\x00\x04\xb7\xb2\xff\xd0\x87\x00\x0foTrnmM9erqfhqpV"]
+rg = raw(gtp)
+print(rg)
```
Comment (p-l-) [inline scapy/contrib/gtp.uts]: Is this a forgotten debug change?

File: scapy/contrib/openflow.py [hunk 3]
```diff
@@ -549,7 +549,7 @@ class OFPTHello(_ofp_header):
                    ByteEnumField("type", 0, ofp_type),
                    ShortField("len", None),
                    IntField("xid", 0)]
-    overload_fields = {TCP: {"sport": 6653}}
+
```
Comment (p-l-) [inline scapy/contrib/openflow.py]: Is this related to this PR?

File: scapy/contrib/gtp.uts (hunk 2 above)
Comment (gpotter2) [reply scapy/contrib/gtp.uts]: This is a bit unrelated, but that test failed on my Windows machine. To replicate the fact that it was printed, then checked, I added a print

File: scapy/contrib/openflow.py (hunk 3 above)
Comment (gpotter2) [reply scapy/contrib/openflow.py]: To make it short, a few protocols used a bind on only one port (only sport, or only dport).
The other one is left to it's default value: 80. Therefore, the HTTP bind would be chosen rather than theirs.

To fix this, those protocols now bind the 2 ports

**Comment by gpotter2** [comment]
@guedou This is ready for review. p-l- has already reviewed a version (not the latest, but there have only been slight bug fixes since). It appears that `tox` is failing to install on Travis with 3.4 - unrelated.

File: scapy/sessions.py [hunk 4]
```diff
@@ -75,3 +79,63 @@ def _ip_process_packet(self, packet):
...
+
+    def __init__(self, *args):
+        IPSession.__init__(self, *args)
+        # A classic dictionary is unpredictable depending on the Python
+        # version. Let's use our custom KeyOrderedDict instead
+        self.tcp_frags = defaultdict(
+            lambda: defaultdict(lambda: (KeyOrderedDict(), {}))
+        )
+
+    def _process_packet(self, pkt):
```
Comment (guedou) [inline scapy/sessions.py]: docstring needed.

File: scapy/sessions.py [hunk 5]
```diff
@@ -75,3 +79,63 @@ def _ip_process_packet(self, packet):
...
+                return pkt
+            self.tcp_frags[ident][ack][1]["pay_class"] = pay_class
+        else:
+            pay_class = self.tcp_frags[ident][ack][1]["pay_class"]
+        self.tcp_frags[ident][ack][0][seq] = pkt
+        # Reassemble using all previous packets
+        return pay_class.tcp_reassemble(*self.tcp_frags[ident][ack])
+
+    def on_packet_received(self, pkt):
+        # First, defragment IP if necessary
```
Comment (guedou) [inline scapy/sessions.py]: docstring needed.

File: scapy/layers/http.py [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+This will decode HTTP packets using `Content_Length` or chunks,
+and will also decompress the packets when needed.
+Note: on failure, decompression will be ignored.
+
+You can turn auto-decompression/auto-compression off with:
+>>> conf.contribs["http"]["auto_compression"] = True
+"""
+
+# This file is a modified version of the former scapy_http plugin.
+# It was reimplemented for scapy 2.4.3+ using sessions, stream handling..
```
Comment (guedou) [inline scapy/layers/http.py]: Remove the last dot.

File: scapy/layers/http.py [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+    def dispatch_hook(cls, _pkt=None, *args, **kargs):
+        if _pkt and False:
+            # XXX TODO
+            from scapy.contrib.contrib.http2 import H2Frame
+            return H2Frame
+        return cls
+
+    # tcp_reassemble is used by TCPSession in session.py
+    @classmethod
+    def tcp_reassemble(cls, frags, metadata):
```
Comment (guedou) [inline scapy/layers/http.py]: Is it possible to split this function using other functions? This will make it easier to read and review.

Comment (gpotter2) [reply scapy/layers/http.py]: It might be a bit hard. This function supports several decompression method, and though there are different, separating it would probably mean duplicating some code.

I've added quite a bunch of comments, but I could add more if it helps

Comment (gpotter2) [reply scapy/layers/http.py]: I've extended the comments to be as explicit as possible. Hope it helps 😕

**Comment by guedou** [comment]
Do we need to add the pcap files? They are quite big.

File: scapy/layers/http.py (hunk 7 above)
Comment (guedou) [reply scapy/layers/http.py]: It sure helps. Don't you think that the steps can be moved to separate functions?

Comment (gpotter2) [reply scapy/layers/http.py]: I don't think that splitting the code will make it easier :/ It doesn't really add anything, and we'll have to scroll to find each part.

Also I'll have to add some extra logic for the returns..

I'd rather keep it as one if you don't mind :-)

**Comment by gpotter2** [comment]
I can probably reduce the size of the pcaps.

**Comment by gpotter2** [comment]
@guedou I've compressed the bigger pcap (`39.8kb -> 24.8kb`). Compressing the other one (`8k`) didn't result in any gain.

I'm not sure if I can spare more space :/ I would have liked to keep real examples for our HTTP tests.

- The first one contain the smallest chunked image I could find. `HTTP chunks` only happen on files that are big enough :/
- The second one is a compressed payload. We could try to reduce it even more by setting something custom, but I'd rather have a test that uses real payloads.
- The "unecessary" packets contained in the pcaps are packets use in TCP ACK, which are very small. Deleting them doesn't reduce the size, but makes the pcap less "real"

**Comment by gpotter2** [comment]
This test: https://github.com/secdev/scapy/blob/master/test/linux.uts#L11
can now be rewritten using Automaton ATMT-SuperSocket conversion mechanisms. Below is a recompiled HTTP payload (2 packets)

```python
>>> load_layer("http")
>>> a = TCP_client.tcplink("www.google.com", 80, full_packets=True)
>>> a.send(b"HEAD / HTTP/1.0\r\n\r\n")
>>> a.sniff(session=TCPSession)
<Sniffed: TCP:1 UDP:0 ICMP:0 Other:0>
>>> _[0]
<Ether  dst=[...] src=[...] type=IPv4 |<IP  version=4 ihl=5 tos=0x0 len=None id=32140 flags= frag=0 ttl=121 proto=tcp chksum=None src=[...] dst=[...] |<TCP  sport=http dport=54773 seq=1135001053 ack=1005612689 dataofs=5 reserved=0 flags=A window=60720 chksum=None urgptr=0 |<HTTP  |<HTTPResponse  Http_Version='HTTP/1.0' Status_Code='200' Reason_Phrase='OK' Accept_Ranges='none' Cache_Control='private, max-age=0' Content_Type='text/html; charset=ISO-8859-1' Date='Sun, 09 Jun 2019 19:22:39 GMT' Expires='-1' P3P='CP="This is not a P3P policy! See g.co/p3phelp for more info."' Server='gws' Set_Cookie='NID=185=kq_kBczt9GkdFomd4pTsi8yue5UIAoODedJ8g7DHFYS1a5M2gGWzGRjXJ6uZgKs4Yim-KtnKEhqzg3pFLVngRtcGj0NQaHzEvkbzMyMDnrmUMaEVYSfVLUZPrX0PgNUOxg2n7_HPmULQ27vBSREWyPTu46QToUwZ8OGGRlBX5bM; expires=Mon, 09-Dec-2019 19:22:39 GMT; path=/; domain=.google.com; HttpOnly' Vary='Accept-Encoding' X_Frame_Options='SAMEORIGIN' X_XSS_Protection='0' |>>>>>
>>>
```

File: scapy/layers/http.py (hunk 7 above)
Comment (gpotter2) [reply scapy/layers/http.py]: I've ended up reworking the function entirely

File: scapy/sessions.py [hunk 8]
```diff
@@ -75,3 +81,134 @@ def _ip_process_packet(self, packet):
...
+            return pkt
+            # Otherwise, maybe store stuff in metadata, and return None,
+            # as you need more fragments to build the packet.
+            return None
+
+    A (hard to understand) example can be found in scapy/layers/http.py
+    """
+
+    def __init__(self, *args):
+        IPSession.__init__(self, *args)
```
Comment (p-l-) [inline scapy/sessions.py]:
```suggestion
        super(TCPSession, self).__init__(*args)
```

File: scapy/arch/pcapdnet.py [hunk 9]
```diff
@@ -266,6 +269,7 @@ class L2pcapListenSocket(_L2pcapdnetSocket):
         desc = "read packets at layer 2 using libpcap"
 
         def __init__(self, iface=None, type=ETH_P_ALL, promisc=None, filter=None, monitor=None):  # noqa: E501
+            _L2pcapdnetSocket.__init__(self)
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/arch/pcapdnet.py]: Is this a convention ?

It's not the first time I've been advised to switch this, but I still find it clearer to use a specific class rather `super()`.

I wouldn't mind changing it if required though

[the remark this answers is not in the document]
Comment (p-l-) [reply scapy/arch/pcapdnet.py]: That is a convention. It has two advantages: if someday you change the parent class, you don't have to change it everywhere; if you use multiple inheritance, calling `super()` guaranties that each parent method will be called, and will be called once.

File: scapy/automaton.py [hunk 10]
```diff
@@ -427,7 +460,7 @@ def __init__(self, name, ioevent, automaton):
         self.automaton = automaton
 
     def __call__(self, proto, *args, **kargs):
-        return _ATMT_supersocket(self.name, self.ioevent, self.automaton, proto, args, kargs)  # noqa: E501
+        return _ATMT_supersocket(self.name, self.ioevent, self.automaton, proto, *args, **kargs)  # noqa: E501
```
Comment (p-l-) [inline scapy/automaton.py]:
```suggestion
        return _ATMT_supersocket(self.name, self.ioevent, self.automaton, proto,
                                 *args, **kargs)
```

File: test/regression.uts [hunk 11]
```diff
@@ -7062,6 +7082,134 @@ assert len(records) == 24
...
...f5e6af2e09e3352047bdca8ce32e46d4b2a66ec8ab5812ec0ab702bb025ac0aec096b0156b025ac0aec0ad644a5ac0aec0968e056b01551bb844d03c7dc8f87e63a65ba7cbc131271b59a7f17118f5e9ef63b420d0f5ce95e08876155c314b7855bc2a9f793f43fd27a906956b696d479abd18feca7d3fab33741a6f127bfd21bf0e3e23e4f52ce99d83b15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762ae2a1810c0107a838080451505053e8ba6cdf6a0553e29f0feacc0cdd95a7c9ce207bb6fb9c886ab247aa5d3f956235304e57c15c03f88a66ab37b3b13f4488f7b950ed13fc412e9fcbba9c552104a3c50d7f03439aacdd89a88721c5ee72e1aec67ad25f2c3344dc654646f06047ebcd5e4c5381a9020f9b9519897236a595327604b5815d812d602ad604b5815d815ac894b5815d812d1c0ad602ad645291ea707a574481f0c9f10f9f7ce8343978f18ef1b3c676be9fc3cc48e52dff005a13335d5ae18a5bc2abe18a49a548a252f248c151475249a01928c49341205bd7740d223d2b4c8ad56864fb53b8fda73d4ff019d569b00c5011f9bb3c70e11498e6433762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8abb15762aec55d8ab
```
Comment (p-l-) [inline test/regression.uts]: Maybe base64 would be better here?

**Comment by p-l-** [comment]
@guedou your call.

**Comment by gpotter2** [comment]
A last small fix & one improvement. This now works properly (it will also follow chunks...):
```python
load_layer("http")
req = HTTPRequest(Accept_Encoding=b'gzip, deflate', Cache_Control=b'no-cache', Connection=b'keep-alive', Host=b'www.secdev.org', Pragma=b'no-cache')
a = TCP_client.tcplink(HTTP, "www.secdev.org", 80)
a.send(req)
payload = a.recv()
open("www.secdev.org.html", "wb").write(payload.load)
```
feel free to try it out on your platform 👍 I only tested the automatons on Windows, but they should work as well elsewhere.

File: test/linux.uts [hunk 12]
```diff
@@ -8,55 +8,6 @@
...
-def handler(signum, stack_frame):
-    raise Exception("Timer expired !")
-
-tmp = signal.signal(signal.SIGALRM, handler)
-
-SECDEV_IP4 = "203.178.141.194"
-IPTABLE_RULE = "iptables -%c INPUT -s %s -p tcp --sport 80 -j DROP"
-
-# Drop packets from SECDEV_IP4
-assert(os.system(IPTABLE_RULE % ('A', SECDEV_IP4)) == 0)
```
Comment (gpotter2) [inline test/linux.uts]: @guedou Hi, could you explain what those do ?
The tests pass on my Linux machine, but fail on Travis :/ If I read this correctly the rule discarded HTTP packets ?!

**Comment by gpotter2** [comment]
Another pretty good demo is available:
```python
load_layer("http")
http_request("www.google.com", display=True)
```
whereas in wirehark:
http_tcp

File: test/linux.uts (hunk 12 above)
Comment (guedou) [reply test/linux.uts]: It drops packets received from the IP address 203.178.141.194 to the TCP port 80. This firewall rule prevents the kernel from receiving the TCP segment and answering with a RST as the port is closed.

Comment (gpotter2) [reply test/linux.uts]: Any idea why the test would be failing ?

Comment (guedou) [reply test/linux.uts]: Is this test disable on current master? I did not see any issue with this test.

Comment (gpotter2) [reply test/linux.uts]: I found out what the issue was: `SECDEV_IPV4` is the ip of `www.kame.org`.

**Comment by gpotter2** [comment]
@guedou Tests now pass. this is mergeable

**Comment by guedou** [comment]
@gpotter2 do you want me to squash the commits or keep them?

**Comment by gpotter2** [comment]
Squashing will be better. There has been some stuff that was reverted

**Comment by p-l-** [comment]
@guedou this is OK for me.

File: scapy/sessions.py [hunk 13]
```diff
@@ -75,3 +81,138 @@ def _ip_process_packet(self, packet):
     def on_packet_received(self, pkt):
         pkt = self._ip_process_packet(pkt)
         DefaultSession.on_packet_received(self, pkt)
+
+
+class StringBuffer(object):
```
Comment (guedou) [inline scapy/sessions.py]: Could you add a docstring?

File: test/regression.uts [hunk 14]
```diff
@@ -2280,8 +2349,8 @@ a.nh == 6 and a.plen == 20 and isinstance(a.payload, TCP) and a.payload.chksum =
 raw(IPv6()/TCP()/Raw(load="somedata")) == b'`\x00\x00\x00\x00\x1c\x06@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x14\x00P\x00\x00\x00\x00\x00\x00\x00\x00P\x02 \x00\xd5\xdd\x00\x00somedata'
 
 = IPv6 Class with TCP and TCP data - dissection
-a=IPv6(raw(IPv6()/TCP()/Raw(load="somedata")))
-a.nh == 6 and a.plen == 28 and isinstance(a.payload, TCP) and a.payload.chksum == 0xd5dd and isinstance(a.payload.payload, Raw) and a[Raw].load == b"somedata"
+a=IPv6(raw(IPv6()/TCP(dport=1234, sport=1234)/Raw(load="somedata")))
```
Comment (guedou) [inline test/regression.uts]: Why did you change this?

Comment (gpotter2) [reply test/regression.uts]: `TCP` has port 80 by default as `sport`.
This would have tried to parse HTTP

File: scapy/sessions.py (hunk 13 above)
Comment (gpotter2) [reply scapy/sessions.py]: Done

File: test/regression.uts (hunk 14 above)
Comment (guedou) [reply test/regression.uts]: Understood!

**Comment by gpotter2** [comment]
I had actually been so focused on HTTP dissection that I overlooked if the build was always correct.
Turns out there were a few issues (corrected in the last commit) with `Unknown_Headers`. This is now fixed.
As both building and dissection are complete, this is (actually) now ready.

Sorry for the last-minute commits

**Comment by gpotter2** [comment]
@guedou @p-l- This should be mergeable.
The Scapy-http maintainer said we would work deprecationnout once this has been merged, to make sure it's ready for 2.4.3 :-)

**Comment by guedou** [comment]
Should I squash this PR?

**Comment by guedou** [comment]
I will try to release v2.4.3rc3 by Wednesday.
